block/src/Pos/Block/Logic/Header.hs

```diff
@@ -255,17 +255,18 @@ getHeadersFromManyTo ::
        , HasConfiguration
        , HasBlockConfiguration
        )
-    => NonEmpty HeaderHash -- ^ Checkpoints; not guaranteed to be
+    => (HeaderHash -> m (Maybe BlockHeader))
```

Comment:
This `getHeader` may or may not do something different than `DB.getHeader`. The caller decides.

Reply:
The only place it is used now is `Logic/Full` and the term is `DB.getHeadersFromManyTo DB.getHeader  checkpoints start`, so this change is meaningless.

Once again: the function has its contract, meaning that input arguments satisfy some conditions, as output arguments (+ effects) also do. I usually think of it in Hoare triple model, but anyway. So, typeclass, among other things, guarantees some properties about its methods. For example, `DB.getHeader` searches rocks for header with given hash and returns the header if it exists. `getHeadersFromManyTo` needs `getHeader` argument to be the function which works like `DB.getHeader` in common sense, otherwise `getHeadersFromManyTo` contract is violated -- comment doesn't say anything about `getHeader` and about result if `getHeader` is doing "something different". Functions should make sense.

Reply:
> The only place it is used now is Logic/Full and the term is DB.getHeadersFromManyTo DB.getHeader checkpoints start, so this change is meaningless.

That's quite short-sighted. It's open for further use, for instance in benchmarking and testing.

> the function has its contract

It still does.

> typeclass, among other things, guarantees some properties about its methods

It guarantees no more than comments on a function parameter do.

> DB.getHeader searches rocks for header with given hash and returns the header if it exists.

This is exactly the problem: there are legit cases where I don't want to use rocks. So I have to define a new monad and make a new instance. I'd rather just swap out a function parameter.

> Functions should make sense.

Ok, that's not controversial...

Reply:
> That's quite short-sighted. It's open for further use, for instance in benchmarking and testing.

Maybe. The point is: this optimization is not needed, because "checkpoints" is a network protocol term, function from current blockchain state, whatever. You may want to use any other set instead of checkpoints, but not in the main application logic, unless we modify the protocol. Anyway, it's not critical.

> It still does.
> It guarantees no more than comments on a function parameter do.

It doesn't say, that `getHeadersFromManyTo` can explode if `getHeader x` returns header for, say, parent for x. And it can, in fact, explode -- with `error`, for example. Nobody guarantees the opposite. Second, you seem to deny the difference between typeclasses and arbitrary functions. The main point is that our codebase has exactly N instances of `MonadDBRead`, also single implementation of `getHeader` for any valid `MonadDBRead`, thus we have about ~N+1 places to make a bug, and we make sure that getting header works in the same way everywhere. If `getHeadersFromManyTo` fails, it's easy for me to debug it, because every usage defines a static set of constraints, so i need to inspect as little code as possible. 

I understand the "but i want function F to be reusable in other non-standard ways for testing and benchmarking", and it's completely valid. What you might want to do instead is to decrease the potential error by passing something different rather than `getHeader` function. Say, if you want to be able to do some effects after/before `getHeader` is called (so `getHeader' = a >> getHeader >>= (\aRes -> b $> aRes)`, you can pass `a` and `b`). 

>So I have to define a new monad and make a new instance. I'd rather just swap out a function parameter.

Exactly, that's the point. This is how our code is structured now. The approach has both advantages and disadvantages. We'll most likely reimplement it sooner or later, according to https://github.com/input-output-hk/cardano-sl/pull/1894. It doesn't make any sense to adopt different approaches in different parts of code for no good reason: it may not be the optimal solution plus it's different from the general one so it complicates code understanding.

Reply:
> > So I have to define a new monad and make a new instance. I'd rather just swap out a function parameter.
>
> Exactly, that's the point. This is how our code is structured now.

It's not. Or do I misunderstand your phrasing? As things stand now in `develop`/`master`, I can't just swap out a function parameter; I have to go via typeclasses. This doesn't add any safety, it just makes the program less composable / harder to use.

>  It doesn't make any sense to adopt different approaches in different parts of code for no good reason: it may not be the optimal solution plus it's different from the general one so it complicates code understanding.

We have now a diffusion layer, and a logic layer. We believe it makes very good sense to adopt a different approach in the diffusion layer, and so we've done just that. We also believe it makes very good sense to adopt that same different approach in the logic layer and indeed everywhere else in cardano-sl, but we've been unable to come to an agreement on that, so it is what it is. Yes it's unfortunate that the logic layer does not follow the same style, but at the very least we have a clear demarcation point between styles.

Reply:
> This getHeader may or may not do something different than DB.getHeader. The caller decides.

Are there any cases when it doesn't something different? If not, then I don't understand why this overcomplication is needed. We can also pass `logDebug`, `sformat`, `getTipHeader`, etc. as arguments, but there is no reason to do it.

> Exactly, that's the point. This is how our code is structured now. The approach has both advantages and disadvantages. We'll most likely reimplement it sooner or later, according to #1894. It doesn't make any sense to adopt different approaches in different parts of code for no good reason: it may not be the optimal solution plus it's different from the general one so it complicates code understanding.

It doesn't really matter now which approach is used. The whole code will be refactored according to `docs/monads.md` at some point (not far in the future). So it's ok to live with different approaches for a while.

>  but we've been unable to come to an agreement on that, so it is what it is.

Well, eventually we've come to an agreement which is currently documented in `docs/monads.md`. My understanding is that consensus is reached.

Reply:
> My understanding is that consensus is reached.

Consensus seems to have been reached among Serokell folks. I opt to use a different, simpler pattern in the diffusion layer.

> It doesn't really matter now which approach is used. The whole code will be refactored according to docs/monads.md at some point (not far in the future). So it's ok to live with different approaches for a while.

The diffusion layer will not be refactored to use that approach.

> Are there any cases when it doesn't something different? If not, then I don't understand why this overcomplication is needed.

Funny, I see it as a simplification rather than a complication. There are indeed cases when it's something different: for instance a benchmark or test case which does not use a database.

I understand that in some sense these are equivalent: the only difference is *how* a different function is swapped in. I'd rather not do it via typeclasses, because it's cumbersome. What's more, it's dishonest: the *monad* in which the result appears really shouldn't determine how a header is retrieved from a hash.

Reply:
> Funny, I see it as a simplification rather than a complication.

Now there are two ways to retrieve a header from a hash: one is to use `getHeader` argument and another one is to use global `DB.getHeader`. Now we have to carefully check that only `getHeader` is used in this function and its subfunctions, not `DB.getHeader`.
🤷‍♂️

Reply:
I don't see a problem. Using the global `DB.getHeader` *within this function* will be a *type error*! Because the required typeclass constraint won't be there. Passing `DB.getHeader` *to this function* will be fine, and a legitimate thing to do.

It's *trivial* to check that this function doesn't use `DB.getHeader` internally, and as I said above, it's even a type error.

Reply:
> Using the global DB.getHeader within this function will be a type error!

No, it won't. Just add `DB.` to one usage of `getHeader` and it will compile. Moreover, there even won't be a warning, because there is also a usage of `getHeader` argument.

> Because the required typeclass constraint won't be there.

It's wrong, you didn't change constraints of this function at all.

> It's trivial to check that this function doesn't use `DB.getHeader` internally

It might be trivial now, but it's a still a manual check, which is always error-prone. But if someone adds `DB.getHeader` usage later, there are high chances that nobody will notice that.

Reply:
Oops, sorry about that, I forgot to remove the `MonadDBRead` constraint.

Reply:
But i don't want to have `getHeader` which will have anything except for getting header from DB. That's what typeclass is for -- it's a convenient way to say "this methods will do X" and document it. Going from typeclass approach to passing everything as an argument just widens the space for a mistake, plus makes it more complicated to guarantee consistency and pre-conditions for arguments. 

Moreover, it's a truly useless refactoring -- you only lower the guarantees for no profit -- there are 0 places the function is used with something different from `DB.getHeader`.

Reply:
> But i don't want to have getHeader which will have anything except for getting header from DB. That's what typeclass is for -- it's a convenient way to say "this methods will do X" and document it. 

That's also what haddock is for, and it can be done for function arguments. *Nothing* stops the programmer from giving a typeclass method which doesn't do what it's supposed to. Nor does anything stop them from giving an argument to a higher-order function which breaks a contract. No safety/correctness is lost here.

> Going from typeclass approach to passing everything as an argument just widens the space for a mistake

It doesn't. It just changes the locus of the mistake.

> plus makes it more complicated to guarantee consistency and pre-conditions for arguments.

It doesn't. With typeclasses: you have N instances to check. With higher-order functions: you have N calls to check.

> Moreover, it's a truly useless refactoring -- you only lower the guarantees for no profit -- there are 0 places the function is used with something different from DB.getHeader.

I do *not* lower any guarantees. The profit is more modularity.

Reply:
> Don't you think it's silly that the monad m alone determines how a header is produced from a header hash? What does the monad have to do with it? Surely there is more than one way to get a header from a header hash in IO.

There are different approaches to monadic effects. We discussed them earlier. They all make sense and have different advantages and disadvantages. In some approaches monad `m` determines the implementation of monadic actions. It makes sense and has advantages as described in #1894. I don't want to get back to this topic.

Reply:
If you eliminate `MonadDBRead` constraint, the idea behind this change will become clear to me.
This refactoring doesn't look related to CSL-2089, so I don't understand why this change was done, and maybe it will be reverted in future, but it should be discussed as part of CBR-52, not this issue.

Reply:
> This refactoring doesn't look related to CSL-2089, so I don't understand why this change was done, and maybe it will be reverted in future, but it should be discussed as part of CBR-52, not this issue.

I checked and yes, it's not necessary for this patch, so I'll put it back to the way it was.